torchgen/gen.py

```diff
@@ -1399,6 +1399,165 @@ def get_native_function_declarations(
     return declarations
 
 
+# Return native function definitions grouped by dispatch key and custom namespace.
+# Used in RegisterDispatchKey.cpp and etc.
+def get_native_function_definitions(
+    *,
+    grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
+    dispatch_key: DispatchKey,
+    backend_idx: BackendIndex,
+    selector: SelectiveBuilder,
+    rocm: bool,
+    skip_dispatcher_op_registration: bool,
+    gen_dispatch_helpers: bool,
+) -> List[str]:
+    definitions: List[str] = []
+    ns_definitions: Dict[str, List[str]] = defaultdict(list)
+    anonymous_definitions: Dict[str, List[str]] = defaultdict(list)
+    registrations: Dict[str, Dict[str, List[str]]] = defaultdict(dict)
+    newline = "\n"
+    ns_gen = dest.RegisterDispatchKey(
+        backend_idx,
+        Target.NAMESPACED_DEFINITION,
+        selector,
+        rocm=rocm,
+        class_method_name=None,
+        skip_dispatcher_op_registration=skip_dispatcher_op_registration,
+    )
+    anonymous_gen = dest.RegisterDispatchKey(
+        backend_idx,
+        Target.ANONYMOUS_DEFINITION,
+        selector,
+        rocm=rocm,
+        class_method_name=None,
+        skip_dispatcher_op_registration=skip_dispatcher_op_registration,
+    )
+    reg_gen = dest.RegisterDispatchKey(
+        backend_idx,
+        Target.REGISTRATION,
+        selector,
+        rocm=rocm,
+        class_method_name=None,
+        skip_dispatcher_op_registration=skip_dispatcher_op_registration,
+    )
+    for f in grouped_native_functions:
+        backend_metadata = backend_idx.get_kernel(f)
+        kernel_namespace = (
+            backend_metadata.cpp_namespace
+            if backend_metadata
+            else DEFAULT_KERNEL_NAMESPACE
+        ).replace("::native", "")
```

Comment:
what's going on here? (I know that on the mobile side we'd rather not have the `native` bit show up in the namespace, but how does this work on the core side?)

I think it would be cleaner to arrange for the correct namespace to already be set as a field on the `NativeFunction` or `BackendMetadata` somewhere, so you don't need to do last-minute string parsing.

Reply:
it's just for the `at::<dispatch_key>` namespace, it's declared in `<DispatchKey>Functions_inl.h`. Example: https://fburl.com/67fm7say